pkg/policy/manager.go

```diff
@@ -0,0 +1,854 @@
+		}
+		s, _ := status.New(codes.InvalidArgument, "failed to compile the provided policy").WithDetails(message)
+		return nil, s.Err()
+	}
+
+	response, err := m.esClient.Bulk(ctx, &esutil.BulkRequest{
+		Index:   m.policiesAlias(),
+		Refresh: m.esConfig.Refresh.String(),
+		Items: []*esutil.BulkRequestItem{
+			{
+				Operation:  esutil.BULK_CREATE,
+				Message:    policy,
+				DocumentId: policyId,
+				Join: &esutil.EsJoin{
+					Field: "join",
+					Name:  "policy",
+				},
+			},
+			{
+				Operation:  esutil.BULK_CREATE,
+				Message:    policyVersion,
+				DocumentId: policyVersionId,
+				Join: &esutil.EsJoin{
+					Parent: policyId,
+					Field:  "join",
+					Name:   "version",
+				},
+			},
+		},
+	})
+
+	if err != nil {
+		return nil, createError(log, "error creating policy", err)
+	}
+
+	var bulkErrors []error
+	for _, item := range response.Items {
+		if item.Create.Error != nil {
+			fmt.Printf("wtaf %v", item.Create.Error)
+			bulkErrors = append(bulkErrors, fmt.Errorf("error creating policy [%d] %s: %s", item.Create.Status, item.Create.Error.Type, item.Create.Error.Reason))
+		}
+	}
+
+	if len(bulkErrors) > 0 {
+		return nil, createError(log, "failed to create policy", fmt.Errorf("bulk creation errors: %v", bulkErrors))
+	}
+
+	//policy := &pb.Policy{
+	//	Id:      uuid.New().String(),
+	//	Policy:  policyEntity,
+	//	Created: currentTime,
+	//	Updated: currentTime,
+	//}
+	//str, err := protojson.MarshalOptions{EmitUnpopulated: true}.Marshal(proto.MessageV2(policy))
+	//if err != nil {
+	//	return nil, createError(log, fmt.Sprintf("error marshalling %T to json", policy), err)
+	//}
+	//
+	//res, err := m.esClient.Index(
+	//	m.indexManager.AliasName(policiesDocumentKind, ""),
+	//	bytes.NewReader(str),
+	//	m.esClient.Index.WithContext(ctx),
+	//	m.esClient.Index.WithRefresh(m.esConfig.Refresh.String()),
+	//)
+	//if err != nil {
+	//	return nil, createError(log, "error sending request to elasticsearch", err)
+	//}
+	//if res.IsError() {
+	//	return nil, createError(log, "error indexing document in elasticsearch", nil, zap.String("response", res.String()), zap.Int("status", res.StatusCode))
+	//}
+	//
+	policy.Policy = policyVersion
+
+	return policy, nil
+}
+
+func (m *manager) GetPolicy(ctx context.Context, request *pb.GetPolicyRequest) (*pb.Policy, error) {
+	log := m.logger.Named("GetPolicy").With(zap.String("id", request.Id))
+
+	response, err := m.esClient.Get(ctx, &esutil.GetRequest{
+		Index:      m.policiesAlias(),
+		DocumentId: request.Id,
+	})
+
+	if err != nil {
+		return nil, createError(log, "error", err)
+	}
+
+	if !response.Found {
+		log.Debug("policy not found")
+
+		return nil, createErrorWithCode(log, "policy not found", err, codes.NotFound)
+	}
+
+	var policy pb.Policy
+	err = protojson.UnmarshalOptions{DiscardUnknown: true}.Unmarshal(response.Source, &policy)
+	if err != nil {
+		return nil, createError(log, "no", err)
+	}
+
+	response, err = m.esClient.Get(ctx, &esutil.GetRequest{
```

Comment:
we are probably going to need to add the ability to specify routing for `Get` operations within the es client package.

https://www.elastic.co/guide/en/elasticsearch/reference/master/parent-join.html#_parent_join_restrictions